ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala

```diff
@@ -193,4 +208,53 @@ private[platform] object EventsBuffer {
         }
       }
   }
+
+  private[cache] def indexAfter(bufferStartInclusiveSearchResult: SearchResult): Int =
+    bufferStartInclusiveSearchResult match {
+      case InsertionPoint(insertionPoint) => insertionPoint
+      case Found(foundIndex) => foundIndex + 1
+    }
+
+  private[cache] def filterAndChunkSlice[ENTRY, FILTER_RESULT](
+      bufferSlice: Vector[(Offset, ENTRY)],
+      filter: ENTRY => Option[FILTER_RESULT],
+      maxChunkSize: Int,
+  ): Inclusive[(Offset, FILTER_RESULT)] = {
+    val bufferSliceIterator = bufferSlice.iterator
+
+    val filteredSlice = bufferSliceIterator
+      .flatMap { case (offset, entry) => filter(entry).map(offset -> _) }
+      .take(maxChunkSize)
+      .toVector
+
+    Inclusive(filteredSlice, Option.when(bufferSliceIterator.hasNext)(filteredSlice.last._1))
+  }
+
+  private[cache] def lastFilteredChunk[ENTRY, FILTER_RESULT](
+      bufferSlice: Vector[(Offset, ENTRY)],
+      filter: ENTRY => Option[FILTER_RESULT],
+      maxChunkSize: Int,
+  ): LastBufferChunkSuffix[(Offset, FILTER_RESULT)] = {
+    val reversedSliceIterator = bufferSlice.view.reverseIterator
+
+    val lastChunk =
```

Comment:
It would be very nice to make this code a little bit more DRY (the iterator transformations and the Option.when part look very similar)